site/src/contexts/ProxyContext.tsx

```diff
@@ -0,0 +1,211 @@
+import { useQuery } from "@tanstack/react-query"
+import { getApplicationsHost, getWorkspaceProxies } from "api/api"
+import { Region } from "api/typesGenerated"
+import { useDashboard } from "components/Dashboard/DashboardProvider"
+import {
+  createContext,
+  FC,
+  PropsWithChildren,
+  useContext,
+  useState,
+} from "react"
+
+interface ProxyContextValue {
+  proxy: PreferredProxy
+  isLoading: boolean
+  error?: Error | unknown
+  setProxy: (regions: Region[], selectedRegion: Region | undefined) => void
+}
+
+interface PreferredProxy {
+  // SelectedRegion is the region the user has selected.
+  // Do not use the fields 'path_app_url' or 'wildcard_hostname' from this
+  // object. Use the preferred fields.
+  selectedRegion: Region | undefined
+  // PreferredPathAppURL is the URL of the proxy or it is the empty string
+  // to indicate using relative paths. To add a path to this:
+  //  PreferredPathAppURL + "/path/to/app"
+  preferredPathAppURL: string
+  // PreferredWildcardHostname is a hostname that includes a wildcard.
+  preferredWildcardHostname: string
+}
+
+export const ProxyContext = createContext<ProxyContextValue | undefined>(
+  undefined,
+)
+
+/**
+ * ProxyProvider interacts with local storage to indicate the preferred workspace proxy.
+ */
+export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
+  // Try to load the preferred proxy from local storage.
+  let savedProxy = loadPreferredProxy()
+  if (!savedProxy) {
+    savedProxy = getURLs([])
+  }
+
+  // The initial state is no regions and no selected region.
+  const [proxy, setProxy] = useState<PreferredProxy>(savedProxy)
+  const setAndSaveProxy = (
+    regions: Region[],
+    selectedRegion: Region | undefined,
```

Comment:
`setAndSaveProxy` should only ever be called deliberately, when we have a region (proxy?) to set. You can check this in your type for `ProxyRow` in `WorkspaceProxyRow.tsx` - the proxy argument is not optional.
```suggestion
    selectedRegion: Region,
```

Reply:
What do you think about not passing `regions` in as an arg, considering we load it right below?
I left another comment in your`WorkspaceProxyPage/hooks.ts` file -  I think we can get away without that hook, considering we make the same API call in here. Should simplify a fair amount.

Reply:
I'm not sure what you mean not passing in the regions.